packages/dds/merge-tree/src/mergeTree.ts

```diff
@@ -893,69 +907,181 @@ export class MergeTree {
+							ref.callbacks?.afterSlide?.(ref);
 						}
 					}
 				}
+			}
 
-				// TODO:AB#4069: This update might be avoidable by checking if the old segment
-				// had hierarchical refs before sliding using `segment.localRefs?.hierRefCount`.
-				if (currentSlideDestination) {
-					this.blockUpdatePathLengths(
-						currentSlideDestination.parent,
-						TreeMaintenanceSequenceNumber,
-						LocalClientId,
-					);
-				}
+			// TODO:AB#4069: This update might be avoidable by checking if the old segment
+			// had hierarchical refs before sliding using `segment.localRefs?.hierRefCount`.
+			if (currentSlideDestination) {
+				this.blockUpdatePathLengths(
+					currentSlideDestination.parent,
+					TreeMaintenanceSequenceNumber,
+					LocalClientId,
+				);
 			}
 		};
-		const segmentCache = new Map<ISegment, { seg?: ISegment }>();
-		for (const segment of segments) {
-			assert(
-				isRemovedAndAcked(segment),
-				0x2f1 /* slideReferences from a segment which has not been removed and acked */,
-			);
-			if (segment.localRefs === undefined || segment.localRefs.empty) {
-				continue;
+
+		const trySlideSegment = (
+			segment: ISegment,
+			currentSlideDestination: ISegment | undefined,
+			currentSlideIsForward: boolean | undefined,
+			currentSlideGroup: LocalReferenceCollection[],
+			pred: (ref: LocalReferencePosition) => boolean,
+			slidingPreference: SlidingPreference,
+			reassign: (
+				localRefs: LocalReferenceCollection,
+				slideToSegment: ISegment | undefined,
+				slideIsForward: boolean,
+			) => void,
+		) => {
+			// avoid sliding logic if this segment doesn't have any references
+			// with the given sliding preference
+			if (!segment.localRefs || !anyLocalRef(segment.localRefs, pred)) {
+				return;
 			}
-			const slideToSegment = this._getSlideToSegment(segment, segmentCache);
+
+			const slideToSegment = this._getSlideToSegment(
+				segment,
+				slidingPreference,
+				segmentCache,
+			);
 			const slideIsForward =
 				slideToSegment === undefined ? false : slideToSegment.ordinal > segment.ordinal;
 
 			if (
 				slideToSegment !== currentSlideDestination ||
 				slideIsForward !== currentSlideIsForward
 			) {
-				slideGroup();
-				currentSlideGroup = [segment.localRefs];
-				currentSlideDestination = slideToSegment;
-				currentSlideIsForward = slideIsForward;
+				slideGroup(currentSlideDestination, currentSlideIsForward, currentSlideGroup, pred);
+				reassign(segment.localRefs, slideToSegment, slideIsForward);
 			} else {
-				currentSlideGroup.push(segment.localRefs);
+				currentRightSlideGroup.push(segment.localRefs);
 			}
+		};
+
+		const segmentCache = new Map<ISegment, { seg?: ISegment }>();
+		for (const segment of segments) {
+			assert(
+				isRemovedAndAcked(segment),
+				0x2f1 /* slideReferences from a segment which has not been removed and acked */,
+			);
+			if (segment.localRefs === undefined || segment.localRefs.empty) {
+				continue;
+			}
+
+			trySlideSegment(
```

Comment:
current code looks like it interleaves references with mixed preferences when the removed section is on one end of the string. I think I would keep something closer to the original structure of the code and not split `currentSlideGroup` up (splitting the other vars up is correct).